File: solutions/python/dimensionality_reduction_ex1.py
# Chapter 15, Exercise 1: PCA on Clinical Lab Data
# Using the simulated metabolic panel data from the chapter

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

# ---- Simulate metabolic data (same as chapter) ----
np.random.seed(42)
n = 300

Sigma = np.array([
    [1.0, 0.7, 0.5, 0.3,-0.2, 0.1, 0.2,-0.1],
    [0.7, 1.0, 0.4, 0.2,-0.2, 0.1, 0.1,-0.1],
    [0.5, 0.4, 1.0, 0.4,-0.3, 0.1, 0.3,-0.1],
    [0.3, 0.2, 0.4, 1.0,-0.5, 0.1, 0.2, 0.0],
    [-0.2,-0.2,-0.3,-0.5, 1.0,-0.1,-0.1, 0.2],
    [0.1, 0.1, 0.1, 0.1,-0.1, 1.0, 0.1,-0.3],
    [0.2, 0.1, 0.3, 0.2,-0.1, 0.1, 1.0,-0.2],
    [-0.1,-0.1,-0.1, 0.0, 0.2,-0.3,-0.2, 1.0]
])

z = np.random.multivariate_normal(np.zeros(8), Sigma, n)
labels = ["glucose", "hba1c", "triglycerides", "ldl",
          "hdl", "creatinine", "alt", "albumin"]

metabolic = pd.DataFrame({
    "glucose": np.round(z[:, 0] * 30 + 100),
    "hba1c": np.round(z[:, 1] * 1.0 + 5.8, 1),
    "triglycerides": np.round(z[:, 2] * 50 + 150),
    "ldl": np.round(z[:, 3] * 30 + 120),
    "hdl": np.round(z[:, 4] * 12 + 55),
    "creatinine": np.round(z[:, 5] * 0.3 + 1.0, 2),
    "alt": np.round(z[:, 6] * 15 + 30),
    "albumin": np.round(z[:, 7] * 0.4 + 4.0, 1)
})

# ---- (a) PCA on standardised data ----
print("=== Part (a): PCA on Standardised Data ===")
X = metabolic.values
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)
pca = PCA()
scores = pca.fit_transform(X_scaled)

print("Explained variance ratio:", np.round(pca.explained_variance_ratio_, 3))

# ---- (b) Scree plot and component selection ----
print("\n=== Part (b): Scree Plot and Component Selection ===")

var_prop = pca.explained_variance_ratio_
cum_var = np.cumsum(var_prop)

fig, axes = plt.subplots(1, 2, figsize=(12, 5))

# Scree plot
axes[0].bar(range(1, 9), var_prop, color="steelblue", edgecolor="white")
axes[0].set_xlabel("Component")
axes[0].set_ylabel("Proportion of Variance")
axes[0].set_title("Scree Plot")
axes[0].set_xticks(range(1, 9))

# Cumulative variance
axes[1].plot(range(1, 9), cum_var, "o-", color="steelblue", lw=2)
axes[1].axhline(y=0.80, color="firebrick", linestyle="--", label="80% threshold")
axes[1].set_xlabel("Number of Components")
axes[1].set_ylabel("Cumulative Proportion")
axes[1].set_title("Cumulative Variance")
axes[1].set_xticks(range(1, 9))
axes[1].legend()

plt.tight_layout()
plt.savefig("ch15_ex1_scree.png", dpi=150)
plt.show()

n_80 = np.argmax(cum_var >= 0.80) + 1

print(f"80% cumulative variance threshold: {n_80} components")
print(f"Cumulative variance: {np.round(cum_var, 3)}")

# ---- (c) Loadings of first two PCs ----
print("\n=== Part (c): Loadings and Interpretation ===")

for pc_idx in [0, 1]:
    loadings = pca.components_[pc_idx]
    order = np.argsort(np.abs(loadings))[::-1]
    print(f"\nPC{pc_idx+1} loadings (sorted by |loading|):")
    for idx in order:
        print(f"  {labels[idx]:>15s}: {loadings[idx]:+.3f}")

print("\nClinical interpretation:")
print("PC1: Dominated by glucose, HbA1c, triglycerides (positive) and")
print("     HDL (negative). This is a 'metabolic syndrome' axis.")
print("PC2: Dominated by creatinine and albumin (opposite signs).")
print("     This captures a 'renal/hepatic function' dimension.")

# ---- (d) Biplot coloured by diabetes status ----
print("\n=== Part (d): Biplot with Diabetes Status ===")

diabetes = np.where(
    (metabolic["glucose"] > 110) & (metabolic["hba1c"] > 6.2),
    "Diabetic", "Non-diabetic"
)
print(f"Diabetes prevalence: {(diabetes == 'Diabetic').mean():.3f}")

fig, ax = plt.subplots(figsize=(9, 7))

# Plot scores
for label, color in [("Non-diabetic", "steelblue"), ("Diabetic", "firebrick")]:
    mask = diabetes == label
    ax.scatter(scores[mask, 0], scores[mask, 1], c=color, s=20,
               alpha=0.6, label=label)

# Add loading arrows
loadings_2d = pca.components_[:2].T
scale_factor = 3
for i, lab in enumerate(labels):
    ax.annotate("", xy=(loadings_2d[i, 0]*scale_factor,
                          loadings_2d[i, 1]*scale_factor),
                 xytext=(0, 0),
                 arrowprops=dict(arrowstyle="->", color="0.3", lw=1.5))
    ax.text(loadings_2d[i, 0]*scale_factor*1.15,
             loadings_2d[i, 1]*scale_factor*1.15,
             lab, fontsize=8, color="0.2", ha="center")

ax.set_xlabel(f"PC1 ({var_prop[0]:.1%} var)")
ax.set_ylabel(f"PC2 ({var_prop[1]:.1%} var)")
ax.set_title("PCA Biplot Coloured by Diabetes Status")
ax.axhline(0, color="grey", linewidth=0.5)
ax.axvline(0, color="grey", linewidth=0.5)
ax.legend()
plt.tight_layout()
plt.savefig("ch15_ex1_biplot.png", dpi=150)
plt.show()

print("\nDiabetic patients tend to cluster toward higher PC1 values,")
print("aligning with the 'metabolic syndrome' interpretation.")
print("PC1 captures the metabolic dimension along which diabetic")
print("patients differ from non-diabetic patients.")

File: solutions/R/dimensionality_reduction_ex1.R
# Chapter 15, Exercise 1: PCA on Clinical Lab Data
# Using the simulated metabolic panel data from the chapter

library(tidyverse)
library(MASS)

# ---- Simulate metabolic data (same as chapter) ----
set.seed(42)
n <- 300

Sigma <- matrix(c(
  1.0, 0.7, 0.5, 0.3,-0.2, 0.1, 0.2,-0.1,
  0.7, 1.0, 0.4, 0.2,-0.2, 0.1, 0.1,-0.1,
  0.5, 0.4, 1.0, 0.4,-0.3, 0.1, 0.3,-0.1,
  0.3, 0.2, 0.4, 1.0,-0.5, 0.1, 0.2, 0.0,
 -0.2,-0.2,-0.3,-0.5, 1.0,-0.1,-0.1, 0.2,
  0.1, 0.1, 0.1, 0.1,-0.1, 1.0, 0.1,-0.3,
  0.2, 0.1, 0.3, 0.2,-0.1, 0.1, 1.0,-0.2,
 -0.1,-0.1,-0.1, 0.0, 0.2,-0.3,-0.2, 1.0
), nrow = 8)

z <- mvrnorm(n, mu = rep(0, 8), Sigma = Sigma)
metabolic <- tibble(
  glucose       = round(z[,1] * 30 + 100),
  hba1c         = round(z[,2] * 1.0 + 5.8, 1),
  triglycerides = round(z[,3] * 50 + 150),
  ldl           = round(z[,4] * 30 + 120),
  hdl           = round(z[,5] * 12 + 55),
  creatinine    = round(z[,6] * 0.3 + 1.0, 2),
  alt           = round(z[,7] * 15 + 30),
  albumin       = round(z[,8] * 0.4 + 4.0, 1)
)

# ---- (a) Perform PCA on standardised data ----
cat("=== Part (a): PCA on Standardised Data ===\n")
pca_fit <- prcomp(metabolic, scale. = TRUE)
cat("PCA completed. Summary:\n")
print(summary(pca_fit))

# ---- (b) Scree plot and number of components ----
cat("\n=== Part (b): Scree Plot and Component Selection ===\n")

var_prop <- pca_fit$sdev^2 / sum(pca_fit$sdev^2)
cum_var  <- cumsum(var_prop)

par(mfrow = c(1, 2), mar = c(4, 4, 3, 1))

# Scree plot
barplot(var_prop, names.arg = paste0("PC", 1:8), col = "steelblue",
        main = "Scree Plot", ylab = "Proportion of Variance",
        xlab = "Component", ylim = c(0, 0.4))
# Cumulative variance
plot(1:8, cum_var, type = "b", pch = 16, col = "steelblue",
     main = "Cumulative Variance", xlab = "Number of Components",
     ylab = "Cumulative Proportion", ylim = c(0, 1))
abline(h = 0.80, col = "firebrick", lty = 2)
text(6, 0.82, "80% threshold", col = "firebrick", cex = 0.8)

n_80 <- which(cum_var >= 0.80)[1]

cat("80% cumulative variance threshold:", n_80, "components\n")
cat("Cumulative variance:", round(cum_var, 3), "\n")

# ---- (c) Loadings of first two PCs ----
cat("\n=== Part (c): Loadings and Interpretation ===\n")

cat("\nPC1 loadings (sorted by absolute value):\n")
pc1_loadings <- pca_fit$rotation[, 1]
print(round(sort(abs(pc1_loadings), decreasing = TRUE), 3))
cat("Signed loadings:\n")
print(round(pc1_loadings[order(abs(pc1_loadings), decreasing = TRUE)], 3))

cat("\nPC2 loadings (sorted by absolute value):\n")
pc2_loadings <- pca_fit$rotation[, 2]
print(round(sort(abs(pc2_loadings), decreasing = TRUE), 3))
cat("Signed loadings:\n")
print(round(pc2_loadings[order(abs(pc2_loadings), decreasing = TRUE)], 3))

cat("\nClinical interpretation:\n")
cat("PC1: Dominated by glucose, HbA1c, triglycerides (positive) and\n")
cat("     HDL (negative). This is a 'metabolic syndrome' axis.\n")
cat("     Patients scoring high on PC1 tend to have elevated glucose,\n")
cat("     HbA1c, triglycerides and lower HDL.\n\n")
cat("PC2: Dominated by creatinine and albumin (with opposite signs).\n")
cat("     This captures a 'renal/hepatic function' dimension.\n")
cat("     Patients with high creatinine and low albumin (suggesting\n")
cat("     renal impairment or liver dysfunction) score high on PC2.\n")

# ---- (d) Biplot coloured by diabetes status ----
cat("\n=== Part (d): Biplot with Diabetes Status ===\n")

set.seed(99)
metabolic$diabetes <- factor(
  ifelse(metabolic$glucose > 110 & metabolic$hba1c > 6.2,
         "Diabetic", "Non-diabetic")
)

cat("Diabetes prevalence:", mean(metabolic$diabetes == "Diabetic"), "\n")

# Extract PC scores
pc_scores <- as.data.frame(pca_fit$x[, 1:2])
pc_scores$diabetes <- metabolic$diabetes

par(mfrow = c(1, 1))

# Biplot with diabetes colouring
cols <- ifelse(metabolic$diabetes == "Diabetic", "firebrick", "steelblue")
plot(pc_scores$PC1, pc_scores$PC2, col = cols, pch = 16, cex = 0.7,
     xlab = paste0("PC1 (", round(var_prop[1] * 100, 1), "% var)"),
     ylab = paste0("PC2 (", round(var_prop[2] * 100, 1), "% var)"),
     main = "PCA Biplot Coloured by Diabetes Status")

# Add loading arrows
loadings <- pca_fit$rotation[, 1:2]
scale_factor <- 3
for (i in 1:nrow(loadings)) {
  arrows(0, 0, loadings[i, 1] * scale_factor, loadings[i, 2] * scale_factor,
         col = "grey30", length = 0.1, lwd = 1.5)
  text(loadings[i, 1] * scale_factor * 1.15,
       loadings[i, 2] * scale_factor * 1.15,
       rownames(loadings)[i], cex = 0.7, col = "grey20")
}

legend("topright", c("Diabetic", "Non-diabetic"),
       col = c("firebrick", "steelblue"), pch = 16, cex = 0.8)

cat("\nDiabetic patients tend to cluster toward higher PC1 values,\n")
cat("which aligns with the 'metabolic syndrome' interpretation.\n")
cat("This confirms that PC1 captures the metabolic dimension\n")
cat("along which diabetic patients differ from non-diabetic patients.\n")
